x-pack/plugin/ilm/src/test/java/org/elasticsearch/xpack/slm/SLMSnapshotBlockingIntegTests.java

```diff
@@ -0,0 +1,278 @@
+
+        // Create a snapshot and wait for it to be complete (need something that can be deleted)
+        final String completedSnapshotName = executePolicy(policyId);
+        logger.info("--> kicked off snapshot {}", completedSnapshotName);
+        assertBusy(() -> {
+            try {
+                SnapshotsStatusResponse s =
+                    client().admin().cluster().prepareSnapshotStatus(repoId).setSnapshots(completedSnapshotName).get();
+                assertThat("expected a snapshot but none were returned", s.getSnapshots().size(), equalTo(1));
+                SnapshotStatus status = s.getSnapshots().get(0);
+                logger.info("--> waiting for snapshot {} to be completed, got: {}", completedSnapshotName, status.getState());
+                assertThat(status.getState(), equalTo(SnapshotsInProgress.State.SUCCESS));
+            } catch (SnapshotMissingException e) {
+                fail("expected a snapshot with name " + completedSnapshotName + " but it does not exist");
+            }
+        });
+
+        // Take another snapshot, but before doing that, block it from completing
+        logger.info("--> blocking nodes from completing snapshot");
+        blockAllDataNodes(repoId);
+        final String secondSnapName = executePolicy(policyId);
+
+        // Check that the executed snapshot shows up in the SLM output as in_progress
+        assertBusy(() -> {
+            GetSnapshotLifecycleAction.Response getResp =
+                client().execute(GetSnapshotLifecycleAction.INSTANCE, new GetSnapshotLifecycleAction.Request(policyId)).get();
+            logger.info("--> checking for in progress snapshot...");
+
+            assertThat(getResp.getPolicies().size(), greaterThan(0));
+            SnapshotLifecyclePolicyItem item = getResp.getPolicies().get(0);
+            assertNotNull(item.getSnapshotInProgress());
+            SnapshotLifecyclePolicyItem.SnapshotInProgress inProgress = item.getSnapshotInProgress();
+            assertThat(inProgress.getSnapshotId().getName(), equalTo(secondSnapName));
+            assertThat(inProgress.getStartTime(), greaterThan(0L));
+            assertThat(inProgress.getState(), anyOf(equalTo(SnapshotsInProgress.State.INIT), equalTo(SnapshotsInProgress.State.STARTED)));
+            assertNull(inProgress.getFailure());
+        });
+
+        // Run retention every second
+        client().admin().cluster().prepareUpdateSettings()
+            .setTransientSettings(Settings.builder()
+                .put(LifecycleSettings.SLM_RETENTION_SCHEDULE, "*/1 * * * * ?")
+                .build())
+            .get();
+        // Guarantee that retention gets a chance to run before unblocking, I know sleeps are not
+        // ideal, but we don't currently have a way to force retention to run, so waiting at least
+        // a second is the best we can do for now.
+        Thread.sleep(1500);
```

Comment:
Why do we need this sleep here when we didn't in the REST version? Is this just to make sure that retention tries to run once, which the original didn't?

Reply:
The REST version had an assertBusy around waiting for the snapshot to disappear. In this case, I didn't want to block snapshots, set retention to run, and then unblock snapshots before the first retention run, because then it wouldn't exercise the retry-on-snapshot-running logic in the code.